qualtran/_infra/data_types.py

```diff
@@ -0,0 +1,147 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import abc
+from typing import TypeVar, Union
+
+import attrs
+import sympy
+
+
+class QDType:
+    @property
+    @abc.abstractmethod
+    def num_qubits(self):
+        """Number of qubits required to represent a single instance of this data type."""
+
+
+QDTypeT = TypeVar("QDTypeT", bound=QDType)
+
+
+@attrs.frozen
+class QBit(QDType):
+    """A single qubit. The smallest addressable unit of quantum data."""
+
+    @property
+    def num_qubits(self):
+        return 1
+
+
+@attrs.frozen
+class QAny(QDType):
+    """Opaque bag-of-qbits type. Should be used sparingly"""
```

Comment:
It will be essential to document or codify the consequences of these data types. I.e. what's the different between `QInt` and `QAny`? is it just a different name to aid the programmer? I see you use `QAny` in the definition of SWAP. Was the intention that you can hook up any dtype to a QAny register?

Reply:
Yes, I think we should (although my type check currently doesn't allow) for wiring any register into a QAny register. QInt etc should be used when we know for sure the register is representing some sort of numeric type (like an int) and not just a chunk of bits we are swapping.

Reply:
Be forewarned that you're sortof implementing your own generic type system. We'll have QAny which is compatible with any QDType. Then you'll get a request for QAnyInt which is compatible with any integer variant. Then you'll get a request for arbitrary "quantum generics"